### PR TITLE
Feature/app 3570

### PR DESCRIPTION
**Pull request by brenthoneybone**
Add ticket and conversation methods

### CODE REVIEW

File: src/rest/methods/conversation.test.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+          content: readFileSync(__dirname + '/../../../test/fixtures/1x1.png'),
+          filename: '2x2.png',
+        },
+      ],
+      body: content,
+      userId: USER_ID,
+    })
+
+    expect(result.content.description).toEqual(content)
+    expect(result._embedded.createdBy.id).toEqual(USER_ID)
```
Comment (yannvgn) [inline src/rest/methods/conversation.test.ts]: is it possible to test the file was attached?

File: src/utils/upload.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  files: ReadonlyArray<{
+    readonly content: Buffer
+    readonly filename: string
+  }>,
+  client: IAllthingsRestClient,
+): Promise<ReadonlyArray<string>> {
+  const result = await upload(files, client)
+
+  return result
+    .filter((item): item is IFile => !(item instanceof Error))
```
Comment (yannvgn) [inline src/utils/upload.ts]: is it ok just to filter out errors?

File: src/rest/methods/ticket.test.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            ),
+            filename: '2x2.png',
+          },
+        ],
+        title: 'title',
+      },
+    )
+
+    expect(description).toEqual('description')
+    expect(title).toEqual('title')
```
Comment (yannvgn) [inline src/rest/methods/ticket.test.ts]: is it possible to test the file was attached?

File: src/rest/methods/ticket.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  WAITING_FOR_EXTERNAL = 'waiting-for-external',
+}
+
+export enum ETrafficLightColor {
+  GREEN = 'green',
+  RED = 'red',
+  YELLOW = 'yellow',
+}
+
+export enum Permission {
```
Comment (yannvgn) [inline src/rest/methods/ticket.ts]: Just wondering: is it worth having this enum here? If yes, we have to make sure this gets maintained / updated as soon as a permission is added.

File: src/rest/methods/ticket.test.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  })
+})
+
+describe('uploadFiles()', () => {
```
Comment (yannvgn) [inline src/rest/methods/ticket.test.ts]: could be put into `utils/upload.test.ts`

File: src/rest/methods/conversation.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+export interface IMessage {
+  readonly content: { readonly content?: string; readonly description?: string }
+  readonly createdAt: string
+  readonly id: string
+  readonly internal: boolean
+  readonly read: boolean
+  readonly type: string
+  readonly _embedded?: any
```
Comment (yannvgn) [inline src/rest/methods/conversation.ts]: is this needed? if so, can it be typed?

File: src/rest/methods/conversation.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  readonly _links?: any
```
Comment (yannvgn) [inline src/rest/methods/conversation.ts]: same as above

File: src/rest/methods/property.ts [hunk 8]
```diff
@@ -7,6 +7,7 @@ export interface IProperty {
   readonly name: string
   readonly timezone: string
   readonly readOnly: boolean
+  readonly _embedded: any
```
Comment (yannvgn) [inline src/rest/methods/property.ts]: is this needed? if so, can it be typed?

File: src/utils/upload.ts (hunk 2 above)
Comment (brenthoneybone) [reply src/utils/upload.ts]: @yannvgn in my case it's good to filter them as a ticket should still be created and not fail if an attachment doesn't get uploaded, it should also add any successful uploads to the ticket. I've renamed this function and added a comment to make it more clear

File: src/rest/methods/ticket.ts (hunk 4 above)
Comment (brenthoneybone) [reply src/rest/methods/ticket.ts]: @yannvgn I think it's overkill so I've removed it and just typed it to string

File: src/rest/methods/conversation.ts (hunk 6 above)
Comment (brenthoneybone) [reply src/rest/methods/conversation.ts]: @yannvgn I'm not sure of all the documents that can be included across all different endpoints. It is however needed to access the property when they are included

File: src/rest/methods/conversation.ts (hunk 7 above)
Comment (brenthoneybone) [reply src/rest/methods/conversation.ts]: @yannvgn I'm not sure of all the documents that can be included across all different endpoints. It is however needed to access the property when they are included

File: src/rest/methods/property.ts (hunk 8 above)
Comment (brenthoneybone) [reply src/rest/methods/property.ts]: @yannvgn I'm not sure of all the documents that can be included across all different endpoints. It is however needed to access the property when they are included

File: src/rest/methods/user.ts [hunk 9]
```diff
@@ -376,3 +376,16 @@ export async function userCheckInToUtilisationPeriod(
...
+  Get a user by email
+ */
+
+export type MethodUserGetByEmail = (email: string) => UserResultList
+
+export async function userGetByEmail(
+  client: IAllthingsRestClient,
+  email: string,
+): UserResultList {
+  return client.getUsers(1, 1000, { email })
```
Comment (xeoxx) [inline src/rest/methods/user.ts]: Why 1, 1000? this feels like it should be options.

File: src/utils/upload.ts (hunk 2 above)
Comment (xeoxx) [reply src/utils/upload.ts]: Can you then not swallow the error but instead return a result that has either an object or an error like this:
```
{
  data?: item,
  error?: error,
}
```

Comment (brenthoneybone) [reply src/utils/upload.ts]: @xeoxx I've updated this method to return both the successful and errors as suggested

File: src/rest/methods/user.ts (hunk 9 above)
Comment (brenthoneybone) [reply src/rest/methods/user.ts]: @xeoxx this is updated

**Pull request merged**